OpenRA.Mods.D2k/Traits/Render/WithCrumbleOverlay.cs

```diff
@@ -35,21 +35,32 @@ public class WithCrumbleOverlayInfo : ConditionalTraitInfo, Requires<RenderSprit
 
 	public class WithCrumbleOverlay : ConditionalTrait<WithCrumbleOverlayInfo>
 	{
+		readonly Animation overlay;
+		readonly RenderSprites renderSprites;
+		readonly WithCrumbleOverlayInfo info;
+		readonly AnimationWithOffset anim;
+
 		public WithCrumbleOverlay(ActorInitializer init, WithCrumbleOverlayInfo info)
 			: base(info)
 		{
+			this.info = info;
+
 			if (init.Contains<SkipMakeAnimsInit>())
 				return;
 
-			var rs = init.Self.Trait<RenderSprites>();
+			renderSprites = init.Self.Trait<RenderSprites>();
 
-			var overlay = new Animation(init.World, rs.GetImage(init.Self));
-			var anim = new AnimationWithOffset(overlay, null, () => IsTraitDisabled);
+			overlay = new Animation(init.World, renderSprites.GetImage(init.Self));
+			anim = new AnimationWithOffset(overlay, null, () => IsTraitDisabled);
 
-			// Remove the animation once it is complete
-			overlay.PlayThen(info.Sequence, () => init.World.AddFrameEndTask(w => rs.Remove(anim)));
+			renderSprites.Add(anim, info.Palette, info.IsPlayerPalette);
+		}
 
-			rs.Add(anim, info.Palette, info.IsPlayerPalette);
+		protected override void TraitEnabled(Actor self)
+		{
+			// Remove the animation once it is complete
+			if (overlay != null && renderSprites != null && anim != null)
```

Comment:
These checks are guaranteed to be true, so there is no need to check them.

Reply:
They aren't. Remove them and the shellmap will crash, any pre-placed building will.

Reply:
I probably can replace this with

```csharp
if (init.Contains<SkipMakeAnimsInit>())
	return;
```

Reply:
Ah, I see. It will be sufficient to check just `overlay` since either all are null or none.